<img src="img/wind-turbine.png" style="float: left; margin: 20px; height: 55px">

# __<p style="color:lightgreen">WHY SWITCH TO WIND TURBINES ?</p>__
#### __<p style="color:lightgreen">A POTENTIAL TO AMPLIFY YOUR INVESTMENT</p>__  

_Green Power: Haleema Esam, Jubran AlTaitoon, Ameera Salman, Khulood Aloun & Zahra Alkhawaja_

---

### Table of Contents

1. [Introduction](#introduction)
    - [Dictionary](#dictionary-function)
    - [Installation & Importing](#installation--imports)
    - [Functions](#functions)
2. [Exploration](#exploration)
    - [Reading Datasets](#read-wind-turbine--operators-data-sets)
    - [Exploring Wing Turbine Dataset](#exploring-wind-turbine-data)
    - [Exploring Operator Dataset](#exploring-operator-data)
    - [Cleaning Operator Dataset](#cleaning-operator-data)
    - [Filter & Further Cleaning Operator Dataset](#filter-operator-table)
3. [Merging Wind Turbine with Wind Operator](#merging-wind_opr_df-with-wind_turbine_df)
4. [Operators: Wind vs. Hydro](#operators-wind-vs-hydro)


<a id='Introduction'></a>

# <p style="color:lightblue">Introduction</p>
---

#### Dictionary Function

In [1]:
# Function to understand the column headers
def dictionary(data_set_type,term):
    '''
        dictionary(type,term=None) -> None

        Prints the term and the definition.
        type:
            0 -> Wind turbine data set
            1 -> Operator data set
    '''
    term=term.lower()
    if data_set_type==0:
        wind_turbine_dictionary={
            'case_id':'Unique stable identification number.',
            'faa_ors':'Unique identifier for cross-reference to the Federal Aviation Administration (FAA) digital obstacle files.',
            'faa_asn':'Unique identifier for cross-reference to the FAA obstruction evaluation airport airspace analysis dataset.',
            'usgs_pr_id':'Unique identifier for cross-reference to the 2014 USGS turbine dataset.',
            't_state':'State where turbine is located.',
            't_county':'County where turbine is located.',
            't_fips':'State and county fips where turbine is located, based on spatial join of turbine points with US state and county.',
            'p_name':'Name of the wind power project that the turbine is a part of. Project names are typically provided by the developer; some names are identified via other internet resources, and others are created by the authors to differentiate them from previous projects. Values are that were unknown were assigned a name based on the county where the turbine is located.',
            'p_year':'Year that the turbine became operational and began providing power. Note this may differ from the year that construction began.',
            'p_tnum':'Number of turbines in the wind power project.',
            'p_cap':'Cumulative capacity of all turbines in the wind power project in megawatts (MW).',
            't_manu':'Turbine manufacturer - name of the original equipment manufacturer of the turbine.',
            't_model':"Turbine model - manufacturer's model name of each turbine.",
            't_cap':'Turbine rated capacity - stated output power at rated wind speed from manufacturer, ACP, and/or internet resources in kilowatts (kW).',
            't_hh':'Turbine hub height in meters (m).',
            't_rd':'Turbine rotor diameter in meters (m).',
            't_rsa':'Turbine rotor swept area in square meters (m2).',
            't_ttlh':'Turbine total height from ground to tip of a blade at its apex in meters (m).',
            'retrofit':'Indicator of whether the turbine has been partially retrofit after initial construction (e.g., rotor and/or nacelle replacement). 0 indicates no known retrofit. 1 indicates yes known retrofit.',
            'retrofit_year':'Year in which the turbine was partially retrofit.',
            't_conf_atr':'Level of confidence in the turbine attributes.\t\n1—No confidence:\tno attribute data beyond total height and year.\t\n2—Partial confidence:\tincomplete information or substantial conflict between.\t\n3—Full confidence:\tcomplete information, consistent across multiple data sources.',
            't_conf_loc':'Level of confidence in turbine location.\t\n1— No turbine shown in image; image has clouds; imagery older than turbine built date.\t\n2— Partial confidence: image shows a developed pad with concrete base and/or turbine parts on the ground.\n3— Full confidence: image shows an installed turbine.',
            't_img_date':'Date of image used to visually verify turbine location. Note if source of image is NAIP, the month and day were set to 01/01.',
            't_img_srce':'Source of image used to visually verify turbine location.',
            'xlong':'Longitude of the turbine point, in decimal degrees.',
            'ylat':'Latitude of the turbine point, in decimal degrees.',
            'eia_id':'Plant ID from Energy Information Administration (EIA).'
        }
        if term in wind_turbine_dictionary:
            print(term,':',wind_turbine_dictionary[term])
        else:
            print(term, 'not in the data set!\ndataset terms:',wind_turbine_dictionary.keys())
    elif data_set_type==1:
        operator_dictionary={
            'plant id':'EIA Plant Identification number. One to five digit numeric.',
            'combined heat and power plant':'Whether or not the plant is a combined heat & power facility (cogenerator).One character alphanumeric, "Y" or "N"',
            'nuclear unit id':'For nuclear plants only, the unit number .One digit numeric. Nuclear plants are the only type of plants for which data are shown explicitly at the generating unit level.',
            'plant name':'Plant name. Alphanumeric',
            'operator name':'The name of the entity which operates the plant. Alphanumeric.',
            'operator id':'The EIA operator identification number. Five digit numeric, padded with leading zeros.',
            'plant state':'State the facility is located in. Two character alphanumeric (standard state postal codes).',
            'census region':'The census region the state is located in.Three to five character alphanumeric:\t\nNEW:\tNew England\t\nMAT:\tMiddle Atlantic\t\nSAT:\tSouth Atlantic\t\nESC:\tEast South Central\t\nWSC:\tWest South Central\t\nENC:\tEast North Central\t\nWNC:\tWest North Central\t\nMTN:\tMountain\t\nPACC:\tPacific Contiguous (Oregon, Washington, and California)\t\nPACN:\tPacific Non-Contiguous (Alaska and Hawaii)',
            'nerc region':'The North American Electric Reliability Council (NERC) region where the plant is located.Alphanumeric:\t\nNPCC:\tNortheast Power Coordinating Council\t\nMRO:\tMidwest Reliability Organization\nSERC:\tSERC Reliability Corporation\nRFC:\tReliability First Corporation\nSPP:\tSouthwest Power Pool\nTRE:\tTexas Regional Entity\t\nFRCC:\tFlorida Reliability Coordinating Council\nWECC:\tWestern Electricity Coordinating Council',
            'reserved':'- No Entry -',
            'naics code':"The facility's North American Industrial Classification System (NAICS) code.Numeric. For information on the NAICS system, see:http://www.census.gov/epcd/www/naics.html",
            'eia sector number':"Number and Sector Name EIA's internal consolidated NAICS sectors.For internal purposes, EIA consolidates the NAICS categories into seven groups.\nThese are shown below in the Sector Codes and Names table:\n1 Electric Utility:\tTraditional regulated electric utilities\n2 NAICS-22 Non-Cogen:\tIndependent power producers which are not cogenerators\n3 NAICS-22 Cogen:\tIndependent power producers which are cogenerators, but whose primary business purpose is the sale of electricity to the public\n4 Commercial NAICS Non-Cogen: Commercial non-cogeneration facilities that produce electric power, are connected to the gird, and can sell power to the public\n5 Commercial NAICS Cogen:\tCommercial cogeneration facilities that produce electric power, are connected to the grid, and can sell power to the public\n6 Industrial NAICS Non-Cogen:\tIndustrial non-cogeneration facilities that produce electric power, are connected to the gird, and can sell power to the public\n7 Industrial NAICS Cogen:\tIndustrial cogeneration facilities that produce electric power, are connected to the gird, and can sell power to the public",
            'sector name':"Sector Name EIA's internal consolidated NAICS sectors.For internal purposes, EIA consolidates the NAICS categories into seven groups.\nThese are shown below in the Sector Codes and Names table:\n1 Electric Utility:\tTraditional regulated electric utilities\n2 NAICS-22 Non-Cogen:\tIndependent power producers which are not cogenerators\n3 NAICS-22 Cogen:\tIndependent power producers which are cogenerators, but whose primary business purpose is the sale of electricity to the public\n4 Commercial NAICS Non-Cogen: Commercial non-cogeneration facilities that produce electric power, are connected to the gird, and can sell power to the public\n5 Commercial NAICS Cogen:\tCommercial cogeneration facilities that produce electric power, are connected to the grid, and can sell power to the public\n6 Industrial NAICS Non-Cogen:\tIndustrial non-cogeneration facilities that produce electric power, are connected to the gird, and can sell power to the public\n7 Industrial NAICS Cogen:\tIndustrial cogeneration facilities that produce electric power, are connected to the gird, and can sell power to the public",
            'reported prime mover':'Type of prime mover:\nBA:\tEnergy Storage, Battery\nBT:\tTurbines Used in a Binary Cycle. Including those used for geothermal applications\nCA:\tCombined-Cycle -- Steam Part\nCE:\tEnergy Storage, Compressed Air\nCP:\tEnergy Storage, Concentrated Solar Power\nCS:\tCombined-Cycle Single-Shaft Combustion Turbine and Steam Turbine share of single generator\nCT:\tCombined-Cycle Combustion Turbine Part\nES:\tEnergy Storage, Other (Specify on Schedule 9, Comments)\nFC:\tFuel Cell\nFW:\tEnergy Storage, Flywheel\nGT:\tCombustion (Gas) Turbine. Including Jet Engine design\nHA:\tHydrokinetic, Axial Flow Turbine\nHB:\tHydrokinetic, Wave Buoy\nHK:\tHydrokinetic, Other\nHY:\tHydraulic Turbine. Including turbines associated with delivery of water by pipeline.\nIC:\tInternal Combustion (diesel, piston, reciprocating) Engine\nPS:\tEnergy Storage, Reversible Hydraulic Turbine (Pumped Storage)\nOT:\tOther\nST:\tSteam Turbine. Including Nuclear, Geothermal, and Solar Steam (does not include Combined Cycle).\nPV:\tPhotovoltaic, WT: Wind Turbine, Onshore,#WS: Wind Turbine, Offshore',
            'reported fuel type code':'The fuel code reported to EIA.Two or three letter alphanumeric:\nAB:\tAgricultural By-Products\nANT:\tAnthracite Coal\nBFG:\tBlast Furnace Gas\nBIT:\tBituminous Coal\nBLQ:\tBlack Liquor\nDFO:\tDistillate Fuel Oil. Including diesel, No. 1, No. 2, and No. 4 fuel oils.\nGEO:\tGeothermal\nJF:\tJet Fuel\nKER:\tKerosene\nLFG:\tLandfill Gas\nLIG:\tLignite Coal\nMSB:\tBiogenic Municipal Solid Waste\nMSN:\tNon-biogenic Municipal Solid Waste\nMWH:\tElectricity used for energy storage\nNG:\tNatural Gas\nNUC:\tNuclear. Including Uranium, Plutonium, and Thorium.\nOBG:\tOther Biomass Gas. Including digester gas, methane, and other biomass gases.\nOBL:\tOther Biomass Liquids\nOBS:\tOther Biomass Solids\nOG:\tOther Gas\nOTH:\tOther Fuel\nPC:\tPetroleum Coke\nPG:\tGaseous Propane\nPUR:\tPurchased Steam\nRC:\tRefined Coal\nRFO:\tResidual Fuel Oil. Including No. 5 & 6 fuel oils and bunker C fuel oil.\nSC:\tCoal-based Synfuel. Including briquettes, pellets, or extrusions, which are formed by binding materials or processes that recycle materials.\nSGC:\tCoal-Derived Synthesis Gas\nSGP:\tSynthesis Gas from Petroleum Coke\nSLW:\tSludge Waste\nSUB:\tSubbituminous Coal\nSUN:\tSolar\nTDF:\tTire-derived Fuels\nWAT:\tWater at a Conventional Hydroelectric Turbine and water used in Wave Buoy Hydrokinetic Technology, current Hydrokinetic Technology, Tidal Hydrokinetic Technology, and Pumping Energy for Reversible (Pumped Storage) Hydroelectric Turbines.\nWC:\tWaste/Other Coal. Including anthracite culm, bituminous gob, fine coal, lignite waste, waste coal.\nWDL:\tWood Waste Liquids, excluding Black Liquor. Including red liquor, sludge wood, spent sulfite liquor, and other wood-based liquids.\nWDS:\tWood/Wood Waste Solids. Including paper pellets, railroad ties, utility polies, wood chips, bark, and other wood waste solids.\nWH:\tWaste Heat not directly attributed to a fuel source\nWND:\tWind\nWO:\tWaste/Other Oil. Including crude oil, liquid butane, liquid propane, naphtha, oil waste, re-refined moto oil, sludge oil, tar oil, or other petroleum-based liquid wastes.',
            'aer fuel type code':'A partial aggregation of the reported fuel type codes into larger categories used by EIA in, for example, the Annual Energy Review (AER).Two or three letter alphanumeric.\nSee the Fuel Code table (Table 5), below:\nSUN:\tSolar PV and thermal\nCOL:\tCoal\nDFO:\tDistillate Petroleum\nGEO:\tGeothermal\nHPS:\tHydroelectric Pumped Storage\nHYC:\tHydroelectric Conventional\nMLG:\tBiogenic Municipal Solid Waste and Landfill Gas\nNG:\tNatural Gas\nNUC:\tNuclear\nOOG:\tOther Gases\nORW:\tOther Renewables\nOTH:\tOther (including nonbiogenic MSW)\nPC:\tPetroleum Coke\nRFO:\tResidual Petroleum\nWND:\tWind\nWOC:\tWaste Coal\nWOO:\tWaste Oil\nWWW:\tWood and Wood Waste',
            'balancing authority code':'The balancing authority code reported to EIA.\nAEC:\tPowerSouth Energy Cooperative\nAECI:\tAssociated Electric Cooperative, Inc.\nAVA:\tAvista Corporation\nAVRN:\tAvangrid Renewables LLC\nAZPS:\tArizona Public Service Company\nBANC:\tBalancing Authority of Northern California\nBPAT:\tBonneville Power Administration\nCHPD:\tPublic Utility District No. 1 of Chelan County\nCISO:\tCalifornia Independent System Operator\nCPLE:\tDuke Energy Progress East\nCPLW:\tDuke Energy Progress West\nCSTO:\tConstellation Energy Control and Dispatch, LLC\nDEAA:\tArlington Valley, LLC - AVBA\nDOPD:\tPUD No. 1 of Douglas County\nDUK:\tDuke Energy Carolinas\nEEI:\tElectric Energy, Inc.\nEPE:\tEl Paso Electric Company\nERCO:\tElectric Reliability Council of Texas, Inc.\nFMPP:\tFlorida Municipal Power Pool\nFPC:\tProgress Energy Florida\nFPL:\tFlorida Power & Light Company\nGCPD:\tPublic Utility District No. 2 of Grant County, Washington\nGRIF:\tGriffith Energy, LLC\nGRIS:\tGridforce South\nGVL:\tGainesville Regional Utilities\nGWA:\tNaturEner Power Watch, LLC (GWA)\nHECO:\tHawaiian Electric Co Inc\nHGMA:\tNew Harquahala Generating Company, LLC - HGBA\nHST:	City of Homestead\nIID:\tImperial Irrigation District\nIPCO:\tIdaho Power Company\nISNE:\tISO New England Inc.\nJEA:\tJEA\nLDWP:\tLos Angeles Department of Water and Power\nLGEE:\tLG&E and KU Services Company as agent for Louisville Gas and Electric Company and Kentucky Utilities\nMISO:\tMidcontinent Independent Transmission System Operator, Inc.\nNA:\tNo Balancing Authority provided\nNBSO:\tNew Brunswick System Operator\nNEVP:\tNevada Power Company\nNSB:\tNew Smyrna Beach, Utilities Commission of\nNWMT:\tNorthWestern Energy (NWMT)\nNYIS:\tNew York Independent System Operator\nOVEC:\tOhio Valley Electric Corporation\nPACE:\tPacifiCorp - East\nPACW:\tPacifiCorp - West\nPGE:\tPortland General Electric Company\nPJM:\tPJM Interconnection, LLC\nPNM:\tPublic Service Company of New Mexico\nPSCO:\tPublic Service Company of Colorado\nPSEI:\tPuget Sound Energy\nSC:\tSouth Carolina Public Service Authority\nSCEG:\tSouth Carolina Electric & Gas Company\nSCL:\tSeattle City Light\nSEC:\tSeminole Electric Cooperative\nSEPA:\tSoutheastern Power Administration\nSOCO:\tSouthern Company Services, Inc. - Trans\nSPA:\tSouthwestern Power Administration\nSRP:\tSalt River Project\nSWPP:\tSouthwest Power Pool\nTAL:\tCity of Tallahassee\nTEC:\tTampa Electric Company\nTEPC:\tTucson Electric Power Company\nTIDC:\tTurlock Irrigation District\nTPWR:\tCity of Tacoma, Department of Public Utilities, Light Division\nTVA:\tTennessee Valley Authority\nWACM:\tWestern Area Power Administration - Rocky Mountain Region\nWALC:\tWestern Area Power Administration - Desert Southwest Region\nWAUW:\tWestern Area Power Administration UGP West\nWWA:\tNaturEner Wind Watch, LLC\nYAD:\tAlcoa Power Generating, Inc. - Yadkin Division',
            'respondent frequency':"Specifies what time period data has to be reported (i.e. monthly data or annual totals) and how often the power plant reports this data to EIA.\nA:\tThe respondent only provides an annual total(s) for this record via the EIA-923 annual survey form.  Any monthly data in this record is estimated based on the respondent's reported annual total(s) and power plants with similar characteristics to this plant.\nAM:\tThe respondent provides monthly values for this record, but does so once per year via the EIA-923 annual survey form.\nM:\tThe respondent provides monthly values for this record and does so via the EIA-923 monthly survey form.",
            'physical unit label':'The type of physical units fuel consumption is reported in:\nmcf:\tfor gases\nshort tons:\tfor solid\nbarrels:\tfor liquids',
            'total quantity consumed in physical units (consumed for electric generation and useful thermal output)':'Consumption of the fuel type in physical units.Numeric.\nNote: this is the total quantity consumed for both electricity and, in the case of combined heat and power plants, process steam production.',
            'quantity consumed in physical units for electric generation':'Consumption for electric generation of the fuel type in physical units.Numeric.',
            'heat content of fuels (mmbtu per unit)':'Heat content of the fuel in millions of Btus per physical unit.Numeric.',
            'total fuel consumed (mmbtu)':'Total consumption of the fuel in millions of Btus.Numeric\nNote: this is the total quantity consumed for both electricity and, in the case of combined heat and power plants, process steam production.',
            'quantity consumed for electricity (mmbtu)':'Consumption of fuel in millions of Btus for the purpose of generating electricity.These fields are most relevant to combined heat and power plants.\nFor non-combined heat and power plants, the data in these fields and in columns AK through AV (total fuel consumed) will be identical.',
            'electricity net generation (mwh)':'Net generation of electricity in megawatthours (MWh).Numeric.  This is total electrical output net of station service.\nIn the case of combined heat and power plants, this value is intended to include internal consumption of electricity for the purposes of a production process, as well as power put on the grid.',
            'total fuel consumption quantity':'Total consumption of fuel in physical units, year to date.Numeric\nNote: this is the total quantity consumed for both electricity and, in the case of combined heat and power plants, process steam production.',
            'electric fuel consumption quantity':'Total consumption of fuel to produce electricity, in physical units, year to date. Numeric',
            'total fuel consumption mmbtu':'Total consumption of fuel in MMBtus, year to date. Numeric   Note: this is the total quantity consumed for both electricity and, in the case of combined heat and power plants, process steam production.',
            'elec fuel consumption mmbtu':'Consumption of fuel in millions of Btus for the purpose of generating electricity, year to date. This field is most relevant to combined heat and power plants.\nFor non-combined heat and power plants, the data in this field and in column BV (total fuel consumed) will be identical.',
            'net generation (megawatthours)':'Net generation, year to date in megawatthours (MWh).Numeric. This is total electrical output net of station service.\nIn the case of combined heat and power plants, this value is intended to include internal consumption of electricity for the purposes of a production process, as well as power put on the grid.',
            'year':'Calendar Year for Data'
            
        }
        if term in operator_dictionary:
            print(term,':',operator_dictionary[term])
        elif ' - ' in term:
            if term[:term.index(' - ')] in operator_dictionary:
                print(term[:term.index(' - ')],':',operator_dictionary[term[:term.index(' - ')]])
            elif term[term.index(' - ')+3:] in operator_dictionary:
                print(term[term.index(' - ')+3:],':',operator_dictionary[term[term.index(' - ')+3:]])
        else:
            print(term, 'not in the data set!\ndataset terms:',operator_dictionary.keys())
    else:
        print('Data set does not exist! Enter between 0 -1 ')

#### Installation & Imports

In [2]:
#Installing libraries:-
#pip install pandas
#pip install numpy
#pip install plotly
#pip install statsmodels

In [3]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go

#### Functions

In [4]:
# Function to clean the column header after reading from Excel 
def clean_header(df):
    '''
        clean_header(df) -> None

        Gets the column names and removes "Unnamed" and "\n" from the header 
    '''
    new_column=[]
    for line1,line2 in df.columns:
        header=''
        sub_header=''
        # replaces Unamed assigned with empty string
        if 'Unnamed' in line1:
            header=''
        else:
            header=line1+' - '
        # replaces \n with empty space
        header+=line2.replace('\n',' ')
        new_column.append(header)
    df.columns=new_column

In [5]:
# Checks the data set for null values
def check(df,show_all=False):
    '''
        check(df,show_all=False) -> None

        Checks the columns of the Dataframe for null and prints the result including percentage of nulls.
        - show_all= default: False - displays only the columns with null values, True displays all coloumns with the number of null values.
    '''
    for col in df.columns:
        if show_all:
            # Check all for Null    
            print(f"{df[col].isnull().sum()} \tNull values in {col[1] if col[0]=='' else col} ({(df[col].dtype)}) with {(df[col].value_counts().shape[0])} unique value(s)")
        else:
            # Only with Null
            null_count=df[col].isnull().sum()
            if null_count>0:
                print(f"{df[col].isnull().sum()} \tNull values ({round(100*null_count/df.shape[0],3)}%) in {col[1] if col[0]=='' else col} ({(df[col].dtype)}) with {(df[col].value_counts().shape[0])} unique value(s)")
        

In [6]:
def get_column_containing(col_list,col_name):
    '''
        get_column_containing(col_list,col_name) -> list()

        Checks the column name list if it contains the name. Then returns a list of column name(s)
    '''
    return [col for col in col_list if col_name in col ]

In [7]:
def convert_to_float(df,columns):
    '''
        convert_to_float(df,columns) -> None

        Converts object columns to float. Replace "." with "Nan" 
  '''
    for col in columns:
        df[col].replace('.', np.nan,inplace=True)
        df[col].astype(float)

### Read Wind Turbine & Operators Data sets

In [8]:
# Reading from wind_turbine_20220114.csv
wind_turbine_df=pd.read_csv('data/wind_turbine_20220114.csv',engine='python',encoding='latin1')
# Reading from Operators data.xlsx - Sheet: Page 1 Generation and Fuel Data
df_dict=pd.read_excel('data/Operators data.xlsx',sheet_name=['Page 1 Generation and Fuel Data'],header=[4,5])
# Extract Page 1 Generation and Fuel Data from datafram dictionary
operator_df=df_dict['Page 1 Generation and Fuel Data']
# Clean the header
clean_header(operator_df)

### Exploring Wind Turbine Data

In [9]:
# Display the first 5 rows
wind_turbine_df.head()

case_id    faa_ors           faa_asn  usgs_pr_id   eia_id t_state  \
0  3072661        NaN               NaN      5149.0  52161.0      CA   
1  3072695        NaN               NaN      5143.0  52161.0      CA   
2  3072704        NaN               NaN      5146.0  52161.0      CA   
3  3063272  19-028134  2014-WTE-4084-OE         NaN      NaN      IA   
4  3053390  19-028015  2015-WTE-6386-OE         NaN      NaN      IA   

       t_county  t_fips                   p_name  p_year  ...     t_rsa  \
0   Kern County    6029                 251 Wind  1987.0  ...       NaN   
1   Kern County    6029                 251 Wind  1987.0  ...       NaN   
2   Kern County    6029                 251 Wind  1987.0  ...       NaN   
3  Story County   19169  30 MW Iowa DG Portfolio  2017.0  ...  12271.85   
4  Boone County   19015  30 MW Iowa DG Portfolio  2017.0  ...  12271.85   

   t_ttlh retrofit retrofit_year  t_conf_atr  t_conf_loc  t_img_date  \
0     NaN        0           NaN           2           3    5/8/2018   
1     NaN        0           NaN           2           3    5/8/2018   
2     NaN        0           NaN           2           3    5/8/2018   
3   150.0        0           NaN           3           3   4/24/2017   
4   150.0        0           NaN           3           3    6/1/2017   

      t_img_srce       xlong       ylat  
0  Digital Globe -118.363762  35.077908  
1  Digital Globe -118.364410  35.077435  
2  Digital Globe -118.364197  35.077644  
3  Digital Globe  -93.430367  42.028233  
4  Digital Globe  -93.700424  41.977608  

[5 rows x 27 columns]

In [10]:
# Display the column names
print(f'Total of {wind_turbine_df.shape[1]} columns\n{wind_turbine_df.columns.to_list()}')

Total of 27 columns
['case_id', 'faa_ors', 'faa_asn', 'usgs_pr_id', 'eia_id', 't_state', 't_county', 't_fips', 'p_name', 'p_year', 'p_tnum', 'p_cap', 't_manu', 't_model', 't_cap', 't_hh', 't_rd', 't_rsa', 't_ttlh', 'retrofit', 'retrofit_year', 't_conf_atr', 't_conf_loc', 't_img_date', 't_img_srce', 'xlong', 'ylat']


In [11]:
# To understand the column names - 0 refers to wind_turbine_df  , 1 refers to operator_df 
dictionary(0,'eia_id')

eia_id : Plant ID from Energy Information Administration (EIA).


In [12]:
# Check for columns with null values (To check all columns send use: check_for_null(wind_turbine_df,True) )
print('Size of data set (Rows , Columns):\n',wind_turbine_df.shape)
print('\nNumber of null values:')
check(wind_turbine_df)

Size of data set (Rows , Columns):
 (70808, 27)

Number of null values:
5792 	Null values (8.18%) in faa_ors (object) with 64999 unique value(s)
5122 	Null values (7.234%) in faa_asn (object) with 65275 unique value(s)
32545 	Null values (45.962%) in usgs_pr_id (float64) with 38241 unique value(s)
5793 	Null values (8.181%) in eia_id (float64) with 1193 unique value(s)
613 	Null values (0.866%) in p_year (float64) with 40 unique value(s)
4482 	Null values (6.33%) in p_cap (float64) with 718 unique value(s)
5640 	Null values (7.965%) in t_manu (object) with 63 unique value(s)
5777 	Null values (8.159%) in t_model (object) with 315 unique value(s)
5480 	Null values (7.739%) in t_cap (float64) with 95 unique value(s)
6180 	Null values (8.728%) in t_hh (float64) with 94 unique value(s)
5934 	Null values (8.38%) in t_rd (float64) with 102 unique value(s)
5934 	Null values (8.38%) in t_rsa (float64) with 103 unique value(s)
6180 	Null values (8.728%) in t_ttlh (float64) with 218 unique value

In [13]:
# Summary Statistics
wind_turbine_df.describe()

case_id    usgs_pr_id        eia_id        t_fips        p_year  \
count  7.080800e+04  38263.000000  65015.000000  70808.000000  70195.000000   
mean   3.058490e+06  27523.587225  57877.751642  32244.494097   2011.650659   
std    3.242377e+04  13564.446593   5984.516487  15498.246596      7.879099   
min    3.000001e+06      1.000000     90.000000   2013.000000   1981.000000   
25%    3.032230e+06  18625.500000  56763.000000  19081.000000   2008.000000   
50%    3.050978e+06  28598.000000  57752.000000  35057.000000   2012.000000   
75%    3.090448e+06  38719.500000  60338.000000  48141.000000   2018.000000   
max    3.118671e+06  49135.000000  65270.000000  72133.000000   2021.000000   

             p_tnum         p_cap         t_cap          t_hh          t_rd  \
count  70808.000000  66326.000000  65328.000000  64628.000000  64874.000000   
mean     104.360552    170.176128   1963.534151     81.054571     95.663717   
std       93.954786    104.478693    717.067912     12.028726     23.424346   
min        1.000000      0.050000     50.000000     19.000000     13.400000   
25%       56.000000     99.000000   1500.000000     80.000000     82.000000   
50%       85.000000    158.000000   2000.000000     80.000000    100.000000   
75%      121.000000    211.220000   2300.000000     87.000000    110.000000   
max      731.000000    525.020000   6000.000000    131.000000    155.000000   

              t_rsa        t_ttlh      retrofit  retrofit_year    t_conf_atr  \
count  64874.000000  64628.000000  70808.000000    5986.000000  70808.000000   
mean    7618.500657    129.050593      0.084538    2018.640327      2.766509   
std     3309.229282     22.185981      0.278196       1.112294      0.597195   
min      141.030000     30.400000      0.000000    2015.000000      1.000000   
25%     5281.020000    121.000000      0.000000    2018.000000      3.000000   
50%     7853.980000    130.100000      0.000000    2019.000000      3.000000   
75%     9503.320000    145.100000      0.000000    2020.000000      3.000000   
max    18869.190000    199.600000      1.000000    2020.000000      3.000000   

         t_conf_loc         xlong          ylat  
count  70808.000000  70808.000000  70808.000000  
mean       2.883784   -100.086457     38.479503  
std        0.462326     11.141025      5.429730  
min        1.000000   -171.713074     13.389381  
25%        3.000000   -103.037155     34.428043  
50%        3.000000    -99.393761     39.052442  
75%        3.000000    -95.202499     42.812810  
max        3.000000    144.722656     66.839905

In [14]:
# Correlation Table
fig = px.imshow(wind_turbine_df.corr(), text_auto=True)
fig.show()

ValueError: could not convert string to float: '19-028134'

### Exploring Operator Data

In [ ]:
# Display the first 5 rows
operator_df.head()

Plant Id Combined Heat And Power Plant Nuclear Unit Id    Plant Name  \
0         1                             N               .    Sand Point   
1         2                             N               .  Bankhead Dam   
2         3                             N               .         Barry   
3         3                             N               .         Barry   
4         3                             N               .         Barry   

                    Operator Name Operator Id Plant State Census Region  \
0  TDX Sand Point Generating, LLC       63560          AK          PACN   
1                Alabama Power Co         195          AL           ESC   
2                Alabama Power Co         195          AL           ESC   
3                Alabama Power Co         195          AL           ESC   
4                Alabama Power Co         195          AL           ESC   

  NERC Region  Reserved  ...  \
0         NaN       NaN  ...   
1        SERC       NaN  ...   
2        SERC       NaN  ...   
3        SERC       NaN  ...   
4        SERC       NaN  ...   

   Electricity Net Generation (MWh) - Netgen September  \
0                                             23.559     
1                                           9187.081     
2                                             254351     
3                                             486059     
4                                          347191.37     

   Electricity Net Generation (MWh) - Netgen October  \
0                                             28.982   
1                                           9136.794   
2                                             220717   
3                                             420338   
4                                          424257.74   

  Electricity Net Generation (MWh) - Netgen November  \
0                                             33.296   
1                                          10514.102   
2                                             248068   
3                                             483657   
4                                          335569.93   

  Electricity Net Generation (MWh) - Netgen December  \
0                                             30.694   
1                                          10935.553   
2                                             227801   
3                                             455550   
4                                           144470.5   

  Year-To-Date - Total Fuel Consumption Quantity  \
0                                              0   
1                                              0   
2                                        1219642   
3                                       52001418   
4                                        1292499   

  Year-To-Date - Electric Fuel Consumption Quantity  \
0                                                 0   
1                                                 0   
2                                           1219642   
3                                          52001418   
4                                           1292499   

  Year-To-Date - Total Fuel Consumption MMBtu  \
0                                        3045   
1                                     1220939   
2                                     1243854   
3                                    53006762   
4                                    26188716   

  Year-To-Date - Elec Fuel Consumption MMBtu  \
0                                       3045   
1                                    1220939   
2                                    1243854   
3                                   53006762   
4                                   26188716   

  Year-To-Date - Net Generation (Megawatthours) Year-To-Date - YEAR  
0                                         347.0                2020  
1                                      139170.0                2020  
2                                     2633517.0                2020  
3                                     50

In [ ]:
# Display the column names - first index of the tupples is column name and the second is the sub-column name
print(f'Total of {operator_df.shape[1]} columns\n{operator_df.columns.to_list()}')

Total of 97 columns
['Plant Id', 'Combined Heat And Power Plant', 'Nuclear Unit Id', 'Plant Name', 'Operator Name', 'Operator Id', 'Plant State', 'Census Region', 'NERC Region', 'Reserved', 'NAICS Code', 'EIA Sector Number', 'Sector Name', 'Reported Prime Mover', 'Reported Fuel Type Code', 'AER Fuel Type Code', 'Balancing Authority Code', 'Respondent Frequency', 'Physical Unit Label', 'Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output) - Quantity January', 'Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output) - Quantity February', 'Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output) - Quantity March', 'Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output) - Quantity April', 'Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output) - Quantity 

In [ ]:
# To understand the column names - 0 refers to wind_turbine_df  , 1 refers to operator_df 
# If the sub-column name does not work try to enter the column name
dictionary(1,'Total Fuel Consumed (MMBtu)')

total fuel consumed (mmbtu) : Total consumption of the fuel in millions of Btus.Numeric
Note: this is the total quantity consumed for both electricity and, in the case of combined heat and power plants, process steam production.


In [ ]:
# Check for columns with null values (To check all columns send use: check_for_null(wind_turbine_df,True) )
print('Size of data set (Rows , Columns):\n',operator_df.shape)
print('\nNumber of null values:')
check(operator_df)

Size of data set (Rows , Columns):
 (15115, 97)

Number of null values:
1 	Null values (0.007%) in Plant Name (object) with 10448 unique value(s)
1 	Null values (0.007%) in Operator Name (object) with 4780 unique value(s)
1 	Null values (0.007%) in Plant State (object) with 51 unique value(s)
565 	Null values (3.738%) in NERC Region (object) with 6 unique value(s)
15115 	Null values (100.0%) in Reserved (float64) with 0 unique value(s)
391 	Null values (2.587%) in Balancing Authority Code (object) with 70 unique value(s)
162 	Null values (1.072%) in Respondent Frequency (object) with 3 unique value(s)
6881 	Null values (45.524%) in Physical Unit Label (object) with 5 unique value(s)


#### Cleaning Operator Data

Dropping 'Reserved' column since all of its values are Null.

In [ ]:
# Dropping 'Reserved' column
operator_df.drop(columns=['Reserved'],axis=1,inplace=True)

Dropping:
- 'Nuclear Unit Id' column since it is used for nuclear plants.
- 'Heat Content Of Fuels (MMBtu Per Unit)' column since it used for fuel plants. (A column with 12 sub-columns)

In [ ]:
# Check the definition of 'Nuclear Unit Id'
dictionary(1,'Nuclear Unit Id')

nuclear unit id : For nuclear plants only, the unit number .One digit numeric. Nuclear plants are the only type of plants for which data are shown explicitly at the generating unit level.


In [ ]:
# Dropping 'Nuclear Unit Id' column
operator_df.drop(columns=['Nuclear Unit Id'],axis=1,inplace=True)

In [ ]:
# Check the definition of 'Heat Content Of Fuels (MMBtu Per Unit)'
dictionary(1,'Heat Content Of Fuels (MMBtu Per Unit)')

heat content of fuels (mmbtu per unit) : Heat content of the fuel in millions of Btus per physical unit.Numeric.


In [ ]:
# Dropping 'Heat Content Of Fuels (MMBtu Per Unit)' column (12 sub-columns)
operator_df.drop(columns=get_column_containing(operator_df.columns,'Heat Content Of Fuels (MMBtu Per Unit)'),axis=1,inplace=True)

#### __Filter & Further Cleaning the Operator Table__:
Extract operator data relating to wind and hydro

In [ ]:
# Filter 'Reported Fuel Type Code' to 'WND' (Wind) and assign it to a new Dataframe (wind_opr_df) and reset index
wind_opr_df=operator_df[operator_df['Reported Fuel Type Code']=='WND']
wind_opr_df.reset_index(drop=True,inplace=True)

# Display first 5 rows
wind_opr_df.head()

Plant Id Combined Heat And Power Plant    Plant Name  \
0         1                             N    Sand Point   
1        90                             N   Snake River   
2       508                             N   Lamar Plant   
3       692                             N  Medicine Bow   
4       944                             N       Geneseo   

                    Operator Name Operator Id Plant State Census Region  \
0  TDX Sand Point Generating, LLC       63560          AK          PACN   
1      Nome Joint Utility Systems       13642          AK          PACN   
2            City of Lamar - (CO)       10633          CO           MTN   
3            SRIV Partnership LLC       62042          WY           MTN   
4          City of Geneseo - (IL)        7095          IL           ENC   

  NERC Region  NAICS Code  EIA Sector Number  ...  \
0         NaN          22                  2  ...   
1         NaN          22                  1  ...   
2        WECC          22                  1  ...   
3        WECC          22                  2  ...   
4         MRO          22                  1  ...   

  Electricity Net Generation (MWh) - Netgen September  \
0                                             23.559    
1                                            161.789    
2                                            901.626    
3                                            765.206    
4                                                420    

  Electricity Net Generation (MWh) - Netgen October  \
0                                            28.982   
1                                           199.034   
2                                           952.676   
3                                          1313.838   
4                                               540   

  Electricity Net Generation (MWh) - Netgen November  \
0                                             33.296   
1                                            228.657   
2                                            1092.82   
3                                           1936.102   
4                                                855   

  Electricity Net Generation (MWh) - Netgen December  \
0                                             30.694   
1                                            210.791   
2                                            1338.32   
3                                           2063.125   
4                                                548   

  Year-To-Date - Total Fuel Consumption Quantity  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   

  Year-To-Date - Electric Fuel Consumption Quantity  \
0                                                 0   
1                                                 0   
2                                                 0   
3                                                 0   
4                                                 0   

  Year-To-Date - Total Fuel Consumption MMBtu  \
0                                        3045   
1                                       20906   
2                                      116085   
3                                      141429   
4                                       51368   

  Year-To-Date - Elec Fuel Consumption MMBtu  \
0                                       3045   
1                                      20906   
2                                     116085   
3                                     141429   
4                                      51368   

  Year-To-Date - Net Generation (Megawatthours) Year-To-Date - YEAR  
0                                         347.0                2020  
1                                        2383.0                2020  
2                                       13232.0                2020  
3                     

In [ ]:
# Filter 'Reported Fuel Type Code' to 'WAT' (Water aka hydro) and assign it to a new Dataframe (hydro_opr_df) and reset index
hydro_opr_df=operator_df[operator_df['Reported Fuel Type Code']=='WAT']
hydro_opr_df.reset_index(drop=True,inplace=True)

# Display first 5 rows
hydro_opr_df.head()

Plant Id Combined Heat And Power Plant          Plant Name  \
0         2                             N        Bankhead Dam   
1         4                             N  Walter Bouldin Dam   
2        11                             N   H Neely Henry Dam   
3        12                             N            Holt Dam   
4        13                             N          Jordan Dam   

      Operator Name Operator Id Plant State Census Region NERC Region  \
0  Alabama Power Co         195          AL           ESC        SERC   
1  Alabama Power Co         195          AL           ESC        SERC   
2  Alabama Power Co         195          AL           ESC        SERC   
3  Alabama Power Co         195          AL           ESC        SERC   
4  Alabama Power Co         195          AL           ESC        SERC   

   NAICS Code  EIA Sector Number  ...  \
0          22                  1  ...   
1          22                  1  ...   
2          22                  1  ...   
3          22                  1  ...   
4          22                  1  ...   

  Electricity Net Generation (MWh) - Netgen September  \
0                                           9187.081    
1                                          36611.875    
2                                           17781.89    
3                                          12304.299    
4                                          38754.735    

  Electricity Net Generation (MWh) - Netgen October  \
0                                          9136.794   
1                                         36411.474   
2                                         17684.558   
3                                         12236.949   
4                                         38542.605   

  Electricity Net Generation (MWh) - Netgen November  \
0                                          10514.102   
1                                          41900.248   
2                                          20350.381   
3                                          14081.583   
4                                          44352.631   

  Electricity Net Generation (MWh) - Netgen December  \
0                                          10935.553   
1                                          43579.795   
2                                          21166.115   
3                                          14646.035   
4                                          46130.482   

  Year-To-Date - Total Fuel Consumption Quantity  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   

  Year-To-Date - Electric Fuel Consumption Quantity  \
0                                                 0   
1                                                 0   
2                                                 0   
3                                                 0   
4                                                 0   

  Year-To-Date - Total Fuel Consumption MMBtu  \
0                                     1220939   
1                                     4865621   
2                                     2363166   
3                                     1635209   
4                                     5150399   

  Year-To-Date - Elec Fuel Consumption MMBtu  \
0                                    1220939   
1                                    4865621   
2                                    2363166   
3                                    1635209   
4                                    5150399   

  Year-To-Date - Net Generation (Megawatthours) Year-To-Date - YEAR  
0                                      139170.0                2020  
1                                      554613.0                2020  
2                                      269368.0                2020  
3                                      186391.0                2020  


##### __Wind operator Table:__

In [ ]:
# Check for columns with null values (To check all columns send use: check_for_null(wind_opr_df,True) )
print('Size of data set (Rows , Columns):\n',wind_opr_df.shape)
print('\nNumber of null values:')
check(wind_opr_df)

Size of data set (Rows , Columns):
 (1243, 83)

Number of null values:
23 	Null values (1.85%) in NERC Region (object) with 6 unique value(s)
15 	Null values (1.207%) in Balancing Authority Code (object) with 35 unique value(s)
8 	Null values (0.644%) in Respondent Frequency (object) with 2 unique value(s)
1243 	Null values (100.0%) in Physical Unit Label (object) with 0 unique value(s)


Dropping 'Physical Unit Label' column since its values are Null

In [ ]:
wind_opr_df.drop(columns=['Physical Unit Label'],axis=1,inplace=True)

C:\Users\jubra\AppData\Local\Temp\ipykernel_12056\3665391784.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Dropping:
- 'Total Fuel Consumption Quantity' column since it does not relate to wind plants.
- 'Electric Fuel Consumption Quantity' column since it does not relate to wind plants.
- 'Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output)' column since it does not relate to wind plants. (12 sub-columns)
- 'Quantity Consumed In Physical Units For Electric Generation' column since it does not relate to wind plants. (A column with 12 sub-columns)

In [ ]:
dictionary(1,'Total Fuel Consumption Quantity')
wind_opr_df['Year-To-Date - Total Fuel Consumption Quantity'].value_counts()

total fuel consumption quantity : Total consumption of fuel in physical units, year to date.Numeric
Note: this is the total quantity consumed for both electricity and, in the case of combined heat and power plants, process steam production.


0    1243
Name: Year-To-Date - Total Fuel Consumption Quantity, dtype: int64

In [ ]:
wind_opr_df.drop(columns=['Year-To-Date - Total Fuel Consumption Quantity'],axis=1,inplace=True)

C:\Users\jubra\AppData\Local\Temp\ipykernel_12056\3786511360.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
dictionary(1,'Electric Fuel Consumption Quantity')
wind_opr_df['Year-To-Date - Electric Fuel Consumption Quantity'].value_counts()

electric fuel consumption quantity : Total consumption of fuel to produce electricity, in physical units, year to date. Numeric


0    1243
Name: Year-To-Date - Electric Fuel Consumption Quantity, dtype: int64

In [ ]:
wind_opr_df.drop(columns=['Year-To-Date - Electric Fuel Consumption Quantity'],axis=1,inplace=True)

C:\Users\jubra\AppData\Local\Temp\ipykernel_12056\939091366.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
dictionary(1,'Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output)')
for col in get_column_containing(wind_opr_df.columns,'Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output)'):
    print(wind_opr_df[col].value_counts())

total quantity consumed in physical units (consumed for electric generation and useful thermal output) : Consumption of the fuel type in physical units.Numeric.
Note: this is the total quantity consumed for both electricity and, in the case of combined heat and power plants, process steam production.
0    1156
.      87
Name: Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output) - Quantity January, dtype: int64
0    1149
.      94
Name: Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output) - Quantity February, dtype: int64
0    1153
.      90
Name: Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output) - Quantity March, dtype: int64
0    1155
.      88
Name: Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output) - Quantity April, dtype: int64
0    1161
.      82
Name: Total Quantity Consumed In 

In [ ]:
dictionary(1,'Quantity Consumed In Physical Units For Electric Generation')
for col in get_column_containing(wind_opr_df.columns,'Quantity Consumed In Physical Units For Electric Generation'):
    print(wind_opr_df[col].value_counts())

quantity consumed in physical units for electric generation : Consumption for electric generation of the fuel type in physical units.Numeric.
0    1156
.      87
Name: Quantity Consumed In Physical Units For Electric Generation - Elec_Quantity January, dtype: int64
0    1149
.      94
Name: Quantity Consumed In Physical Units For Electric Generation - Elec_Quantity February, dtype: int64
0    1153
.      90
Name: Quantity Consumed In Physical Units For Electric Generation - Elec_Quantity March, dtype: int64
0    1155
.      88
Name: Quantity Consumed In Physical Units For Electric Generation - Elec_Quantity April, dtype: int64
0    1161
.      82
Name: Quantity Consumed In Physical Units For Electric Generation - Elec_Quantity May, dtype: int64
0    1164
.      79
Name: Quantity Consumed In Physical Units For Electric Generation - Elec_Quantity June, dtype: int64
0    1165
.      78
Name: Quantity Consumed In Physical Units For Electric Generation - Elec_Quantity July, dtype: int64
0  

In [ ]:
# Dropping columns:
# - 'Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output)' (12 sub-columns)
# - 'Quantity Consumed In Physical Units For Electric Generation' (12 sub-columns)
                          
wind_opr_df.drop(columns=get_column_containing(wind_opr_df.columns,'Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output)'),axis=1,inplace=True)
wind_opr_df.drop(columns=get_column_containing(wind_opr_df.columns,'Quantity Consumed In Physical Units For Electric Generation'),axis=1,inplace=True)

C:\Users\jubra\AppData\Local\Temp\ipykernel_12056\569489568.py:5: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

C:\Users\jubra\AppData\Local\Temp\ipykernel_12056\569489568.py:6: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
print('Size of data set (Rows , Columns):\n',wind_opr_df.shape)

Size of data set (Rows , Columns):
 (1243, 56)


In [ ]:
# Check data type
wind_opr_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1243 entries, 0 to 1242
Data columns (total 56 columns):
 #   Column                                                            Non-Null Count  Dtype  
---  ------                                                            --------------  -----  
 0   Plant Id                                                          1243 non-null   int64  
 1   Combined Heat And Power Plant                                     1243 non-null   object 
 2   Plant Name                                                        1243 non-null   object 
 3   Operator Name                                                     1243 non-null   object 
 4   Operator Id                                                       1243 non-null   object 
 5   Plant State                                                       1243 non-null   object 
 6   Census Region                                                     1243 non-null   object 
 7   NERC Region                      

In [ ]:
# Change data type from object to float
convert_to_float(wind_opr_df,get_column_containing(wind_opr_df.columns,'Total Fuel Consumed (MMBtu)'))
convert_to_float(wind_opr_df,get_column_containing(wind_opr_df.columns,'Quantity Consumed For Electricity (MMBtu)'))
convert_to_float(wind_opr_df,get_column_containing(wind_opr_df.columns,'Electricity Net Generation (MWh)'))

C:\Users\jubra\AppData\Local\Temp\ipykernel_12056\3761527868.py:8: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
# Summary statistics
wind_opr_df.describe()

Plant Id     NAICS Code  EIA Sector Number  \
count   1243.000000    1243.000000        1243.000000   
mean   57312.234916    1076.500402           1.968624   
std     8956.763670   12526.455911           0.687932   
min        1.000000      22.000000           1.000000   
25%    56644.500000      22.000000           2.000000   
50%    57617.000000      22.000000           2.000000   
75%    59493.500000      22.000000           2.000000   
max    99999.000000  325211.000000           7.000000   

       Total Fuel Consumed (MMBtu) - Tot_MMBtu January  \
count                                     1.156000e+03   
mean                                      2.134162e+05   
std                                       2.538814e+05   
min                                       0.000000e+00   
25%                                       1.261325e+04   
50%                                       1.250050e+05   
75%                                       3.331575e+05   
max                                       2.067463e+06   

       Total Fuel Consumed (MMBtu) - Tot_MMBtu February  \
count                                      1.149000e+03   
mean                                       2.222684e+05   
std                                        2.568774e+05   
min                                        0.000000e+00   
25%                                        1.480000e+04   
50%                                        1.334810e+05   
75%                                        3.440140e+05   
max                                        1.918111e+06   

       Total Fuel Consumed (MMBtu) - Tot_MMBtu March  \
count                                   1.153000e+03   
mean                                    2.230897e+05   
std                                     2.529347e+05   
min                                     0.000000e+00   
25%                                     1.564500e+04   
50%                                     1.425490e+05   
75%                                     3.391440e+05   
max                                     1.845901e+06   

       Total Fuel Consumed (MMBtu) - Tot_MMBtu April  \
count                                   1.155000e+03   
mean                                    2.259881e+05   
std                                     2.574561e+05   
min                                     0.000000e+00   
25%                                     1.693900e+04   
50%                                     1.455270e+05   
75%                                     3.481610e+05   
max                                     2.159342e+06   

       Total Fuel Consumed (MMBtu) - Tot_MMBtu May  \
count                                 1.161000e+03   
mean                                  2.144322e+05   
std                                   2.470443e+05   
min                                   0.000000e+00   
25%                                   1.537900e+04   
50%                                   1.402020e+05   
75%                                   3.241360e+05   
max                                   2.049934e+06   

       Total Fuel Consumed (MMBtu) - Tot_MMBtu June  \
count                                  1.164000e+03   
mean                                   2.277091e+05   
std                                    2.791473e+05   
min                                    0.000000e+00   
25%                                    1.601950e+04   
50%                                    1.278235e+05   
75%                                    3.372460e+05   
max                                    2.034932e+06   

       Total Fuel Consumed (MMBtu) - Tot_MMBtu July  ...  \
count                                  1.165000e+03  ...   
mean                                   1.721933e+05  ...   
std                                    2.093712e+05  ...   
min                                    0.000000e+00  ...   
25%                                    1.271400e+04  ...   
50%                                    9.761200e+04  ...   
75%                          

##### __Hydro operator Table:__ 

In [ ]:
# Check for columns with null values (To check all columns send use: check_for_null(wind_opr_df,True) )
print('Size of data set (Rows , Columns):\n',hydro_opr_df.shape)
print('\nNumber of null values:')
check(hydro_opr_df)

Size of data set (Rows , Columns):
 (1477, 83)

Number of null values:
44 	Null values (2.979%) in NERC Region (object) with 6 unique value(s)
38 	Null values (2.573%) in Balancing Authority Code (object) with 45 unique value(s)
6 	Null values (0.406%) in Respondent Frequency (object) with 2 unique value(s)
1435 	Null values (97.156%) in Physical Unit Label (object) with 1 unique value(s)


Dropping 'Physical Unit Label' column since more than 97% of its values are missing

In [ ]:
dictionary(1,"Physical Unit Label")

physical unit label : The type of physical units fuel consumption is reported in:
mcf:	for gases
short tons:	for solid
barrels:	for liquids


In [ ]:
# Drop 'Physical Unit Label'
hydro_opr_df.drop(columns=['Physical Unit Label'],axis=1,inplace=True)

C:\Users\jubra\AppData\Local\Temp\ipykernel_12056\2952157296.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
print('Size of data set (Rows , Columns):\n',hydro_opr_df.shape)

Size of data set (Rows , Columns):
 (1477, 82)


In [ ]:
# Summary Statistis
hydro_opr_df.describe()

Plant Id     NAICS Code  EIA Sector Number  \
count   1477.000000    1477.000000        1477.000000   
mean   14513.930941    3323.263372           1.489506   
std    21403.430214   29663.998413           0.896805   
min        2.000000      22.000000           1.000000   
25%     1516.000000      22.000000           1.000000   
50%     3772.000000      22.000000           1.000000   
75%    10494.000000      22.000000           2.000000   
max    99999.000000  322122.000000           7.000000   

       Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output) - Quantity January  \
count                                        1474.000000                                                                           
mean                                         1291.894844                                                                           
std                                         14355.954403                                                                           
min                                             0.000000                                                                           
25%                                             0.000000                                                                           
50%                                             0.000000                                                                           
75%                                             0.000000                                                                           
max                                        293292.000000                                                                           

       Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output) - Quantity February  \
count                                        1474.000000                                                                            
mean                                         1146.075305                                                                            
std                                         13170.663410                                                                            
min                                             0.000000                                                                            
25%                                             0.000000                                                                            
50%                                             0.000000                                                                            
75%                                             0.000000                                                                            
max                                        279028.000000                                                                            

       Total Quantity Consumed In Physical Units (Consumed For Electric Generation And Useful Thermal Output) - Quantity March  \
count                                        1474.000000                                                                         
mean                                         1152.063094                                                                         
std                                         12005.836079                                                                         
min                                             0.000000                                                                         
25%                                             0.000000                                                                         
50%                                             0.000000                                                                         
75%                                             0.000000                                                                         
max                                        212807.000000              

##### Merging wind_opr_df with wind_turbine_df
- Connect by Plant Id _(wind_turbine_df['eia_id] == wind_opr_df['']['Plant Id'])_

In [ ]:
# Checking plant id values in both dataframes
print('- wind_turbine_df\n')
print(wind_turbine_df['eia_id'].value_counts())
print('\n- wind_opr_df\n')
print(wind_opr_df['Plant Id'].value_counts()) 

- wind_turbine_df

50535.0    432
56291.0    421
56763.0    407
56777.0    355
60619.0    299
          ... 
56376.0      1
56225.0      1
55853.0      1
57257.0      1
62642.0      1
Name: eia_id, Length: 1193, dtype: int64

- wind_opr_df

99999    8
2022     2
58769    1
58833    1
58830    1
        ..
56941    1
56935    1
56934    1
56933    1
57613    1
Name: Plant Id, Length: 1235, dtype: int64


- Plant Id: 99999 does not feature in wind_turbine_df dataframe.

- Plant Id: 2022 is duplicated and featured in the same state but its entries are not. It exists in the wind_turbine_df dataframe but the data variables in both tables are not enough to accurately link plant id 2022 in both tables. Therefore, we will link it by default.

In [ ]:
# Plant Id : 2022 in wind_opr_df
wind_opr_df[wind_opr_df['Plant Id']==2022]

Plant Id Combined Heat And Power Plant Plant Name  \
7      2022                             Y    Willmar   
8      2022                             N    Willmar   

                 Operator Name Operator Id Plant State Census Region  \
7  Willmar Municipal Utilities       20737          MN           WNC   
8  Willmar Municipal Utilities       20737          MN           WNC   

  NERC Region  NAICS Code  EIA Sector Number  ...  \
7         MRO          22                  1  ...   
8         MRO          22                  1  ...   

  Electricity Net Generation (MWh) - Netgen July  \
7                                            NaN   
8                                        395.504   

  Electricity Net Generation (MWh) - Netgen August  \
7                                              NaN   
8                                          529.517   

  Electricity Net Generation (MWh) - Netgen September  \
7                                                NaN    
8                                            558.184    

  Electricity Net Generation (MWh) - Netgen October  \
7                                               NaN   
8                                           568.271   

  Electricity Net Generation (MWh) - Netgen November  \
7                                                NaN   
8                                            670.856   

  Electricity Net Generation (MWh) - Netgen December  \
7                                                NaN   
8                                            612.789   

   Year-To-Date - Total Fuel Consumption MMBtu  \
7                                        33538   
8                                        29258   

   Year-To-Date - Elec Fuel Consumption MMBtu  \
7                                       33538   
8                                       29258   

   Year-To-Date - Net Generation (Megawatthours)  Year-To-Date - YEAR  
7                                       3822.879                 2020  
8                                       3335.121                 2020  

[2 rows x 56 columns]

In [ ]:
# EIA Id (Plant Id) : 2022 in wind_turbine_df
wind_turbine_df[wind_turbine_df['eia_id']==2022]

case_id    faa_ors           faa_asn  usgs_pr_id  eia_id t_state  \
65298  3030660  27-020731  2009-WTE-3378-OE     29363.0  2022.0      MN   
65299  3030661  27-021074  2009-WTE-3379-OE     29364.0  2022.0      MN   

               t_county  t_fips   p_name  p_year  ...    t_rsa  t_ttlh  \
65298  Kandiyohi County   27067  Willmar  2009.0  ...  5026.55   120.1   
65299  Kandiyohi County   27067  Willmar  2009.0  ...  5026.55   120.1   

      retrofit retrofit_year  t_conf_atr  t_conf_loc  t_img_date  \
65298        0           NaN           3           3  11/22/2019   
65299        0           NaN           3           3  11/22/2019   

          t_img_srce      xlong       ylat  
65298  Digital Globe -95.007996  45.156693  
65299  Digital Globe -95.006927  45.159325  

[2 rows x 27 columns]

In [ ]:
# Change dtype of 'Plant Id' in wind_opr_df
wind_opr_df['Plant Id'].astype(float)

0           1.0
1          90.0
2         508.0
3         692.0
4         944.0
         ...   
1238    99999.0
1239    99999.0
1240    99999.0
1241    99999.0
1242    99999.0
Name: Plant Id, Length: 1243, dtype: float64

In [ ]:
# merging both wind_opr_df with wind_turbine_df 
merged_wind_data=wind_turbine_df.merge(right=wind_opr_df,how='left',left_on=['eia_id'],right_on=['Plant Id'])
# Display first 5 rows
merged_wind_data.head()

case_id    faa_ors           faa_asn  usgs_pr_id   eia_id t_state  \
0  3072661        NaN               NaN      5149.0  52161.0      CA   
1  3072695        NaN               NaN      5143.0  52161.0      CA   
2  3072704        NaN               NaN      5146.0  52161.0      CA   
3  3063272  19-028134  2014-WTE-4084-OE         NaN      NaN      IA   
4  3053390  19-028015  2015-WTE-6386-OE         NaN      NaN      IA   

       t_county  t_fips                   p_name  p_year  ...  \
0   Kern County    6029                 251 Wind  1987.0  ...   
1   Kern County    6029                 251 Wind  1987.0  ...   
2   Kern County    6029                 251 Wind  1987.0  ...   
3  Story County   19169  30 MW Iowa DG Portfolio  2017.0  ...   
4  Boone County   19015  30 MW Iowa DG Portfolio  2017.0  ...   

   Electricity Net Generation (MWh) - Netgen July  \
0                                             NaN   
1                                             NaN   
2                                             NaN   
3                                             NaN   
4                                             NaN   

   Electricity Net Generation (MWh) - Netgen August  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   

  Electricity Net Generation (MWh) - Netgen September  \
0                                                NaN    
1                                                NaN    
2                                                NaN    
3                                                NaN    
4                                                NaN    

  Electricity Net Generation (MWh) - Netgen October  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   

   Electricity Net Generation (MWh) - Netgen November  \
0                                                NaN    
1                                                NaN    
2                                                NaN    
3                                                NaN    
4                                                NaN    

   Electricity Net Generation (MWh) - Netgen December  \
0                                                NaN    
1                                                NaN    
2                                                NaN    
3                                                NaN    
4                                                NaN    

   Year-To-Date - Total Fuel Consumption MMBtu  \
0                                          0.0   
1                                          0.0   
2                                          0.0   
3                                          NaN   
4                                          NaN   

   Year-To-Date - Elec Fuel Consumption MMBtu  \
0                                         0.0   
1                                         0.0   
2                                         0.0   
3                                         NaN   
4                                         NaN   

   Year-To-Date - Net Generation (Megawatthours)  Year-To-Date - YEAR  
0                                            0.0               2020.0  
1                                            0.0               2020.0  
2                                            0.0               2020.0  
3                                            NaN                  NaN  
4                                            NaN                  NaN  

[5 rows x 83 columns]

In [ ]:
# Correlation Table
fig = px.imshow(merged_wind_data.corr(), text_auto=True)
fig.show()

C:\Users\jubra\AppData\Local\Temp\ipykernel_12056\2175453109.py:2: FutureWarning:

The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.



# Operators: Wind vs. Hydro

##### __Operators & Plants:__

In [ ]:
# Number of Plants
print(f'Number of plants: {wind_opr_df["Plant Id"].unique().shape[0]} wind operating plants')
print(f'Number of plants: {hydro_opr_df["Plant Id"].unique().shape[0]} hydro operating plants')

# Number of Wind Operators
print(f"Number of wind operators: {wind_opr_df['Operator Id'].unique().shape[0]} ")
print(f"Number of hydro operators: {hydro_opr_df['Operator Id'].unique().shape[0]} ")

Number of plants: 1235 wind operating plants
Number of plants: 1457 hydro operating plants
Number of wind operators: 653 
Number of hydro operators: 469 


In [ ]:
# Number of plants per state (Hover over the states to check for details)
print('Number of wind operating plants per state:')
fig = px.choropleth(locations=wind_opr_df['Plant State'].value_counts().index, locationmode="USA-states",\
                     color=wind_opr_df['Plant State'].value_counts().values, scope="usa",\
                      title='Number of Wind operating plant per state')
fig.show()

print('Number of hydro operating plants per state:')
fig = px.choropleth(locations=hydro_opr_df['Plant State'].value_counts().index, locationmode="USA-states", \
                    color=hydro_opr_df['Plant State'].value_counts().values, scope="usa",\
                      title='Number of Hydro operating plant per state')
fig.show()

Number of wind operating plants per state:


Number of hydro operating plants per state:


##### __Total - Net Generation (Megawatthours) for 2022:__

In [ ]:
print(f"Wind Plants: Total Electrical Output (Megawatthours): {wind_opr_df[['Year-To-Date - Net Generation (Megawatthours)']].sum()[0]}")
print(f"Hydro Plants: Total Electrical Output (Megawatthours): {hydro_opr_df[['Year-To-Date - Net Generation (Megawatthours)']].sum()[0]}")

fig = go.Figure()
fig.add_trace(go.Bar(
    x=['Wind Plants'],
    y=wind_opr_df[['Year-To-Date - Net Generation (Megawatthours)']].sum(),
    name='Wind Operators',
    marker_color='green'
))
fig.add_trace(go.Bar(
    x=['Hydro Plants'],
    y=hydro_opr_df[['Year-To-Date - Net Generation (Megawatthours)']].sum(),
    name='Hydro Operators',
    marker_color='blue'
))
fig.update_layout(title_text='Total Electrical Output',)
# Here we modify the tickangle of the xaxis, resulting in rotated labels.
fig.update_layout(barmode='group', xaxis_tickangle=-45)
fig.show()

Wind Plants: Total Electrical Output (Megawatthours): 337938049.026
Hydro Plants: Total Electrical Output (Megawatthours): 279952689.827
